Add Dimensions controls to Product Collection block

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR introduces `Width` controls to the `Product Collection` block, enabling users to define fixed or flexible width options.

The implementation adds the width configuration in the block inspector. The controls are placed at the `Product Collection` level and affect the block and its children. So width configuration will affect the `Product Template`, the `Pagination` and the `No Results` blocks.

**Why this change is needed**: This feature enhances flexibility in product display customization, allowing users to better manage the presentation of their product. 

https://github.com/user-attachments/assets/b9678179-49ed-45a1-b00e-3807baf07b28

Closes #47884.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a few products.
2. Navigate to Pages > Add New Page and insert a `Product Collection` block.
3. In the block setup dialog, select `create your own` option.
4. Press the settings button to open the inspector and scroll down to the `Dimensions` options.
5. Under the `Width` options, select `Fixed` and add a value (e.g. 400px).
6. Verify that the `Product Template` block reflects this width immediately.
7. Modify the fixed value and verify that the width changes immediately again.
8. Select the `Fill` option. The `Product Collection` block and its children should fill the available width.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

octaedro committed:
Add dimensions to product collection

octaedro committed:
Refactor width-options-control

octaedro committed:
Add product-template changes

octaedro committed:
Fix dimensions

octaedro committed:
Add changelog

octaedro committed:
Remove minHeight

octaedro committed:
Remove className addition

octaedro committed:
Small refactor in ProductTemplate

octaedro committed:
Remove variable

octaedro committed:
Fix lint

octaedro committed:
undo product-template changes

octaedro committed:
Add ProductCollection fixed width

octaedro committed:
Fix set_fixed_width_style param comment

octaedro committed:
Deconstruct WidthOptionsControl props

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx:
```diff
@@ -16,7 +16,6 @@ import {
 } from '@wordpress/block-editor';
 import { Spinner } from '@wordpress/components';
 import { store as coreStore } from '@wordpress/core-data';
-import { ProductCollectionAttributes } from '@woocommerce/blocks/product-collection/types';
```
I think it's in use so should be brought back.

octaedro replied:
Good eyes! Added in the commit [cb8407d](https://github.com/woocommerce/woocommerce/pull/51420/commits/cb8407d66ce45e22bfa6da39ee36cd086f423f48)

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -136,6 +136,45 @@ function ( $html ) {
+		$style = '';
+
+		if ( isset( $fixed_width ) ) {
+			$style .= sprintf( 'width:%dpx;', intval( $fixed_width ) );
```
Control allows you to change units so we should not assume it's `px`. BTW unit is respected in Editor

<img width="286" alt="image" src="https://github.com/user-attachments/assets/eb77d32b-c819-4b9a-ab18-e9da63ebea40">

octaedro replied:
Good call! I fixed this in the commit [1620002](https://github.com/woocommerce/woocommerce/pull/51420/commits/1620002ff43b43b40ef7e02ede420333ed8f7b52)

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/width-options-control.tsx:
```diff
@@ -0,0 +1,88 @@
+	return (
+		<ToolsPanelItem
+			label={ __( 'Width', 'woocommerce' ) }
+			hasValue={ () => true }
```
I suggest using actual conditions for `hasValue` so basically whenever the value is different from a default:

```suggestion
			hasValue={ () => widthType !== WidthOptions.FILL }
			isShownByDefault
```

You can even import the default value from `plugins/woocommerce-blocks/assets/js/blocks/product-collection/constants.ts` for comparison.

If you wanted to make sure option is shown by default, use `isShownByDefault` prop as in the suggestion above.

kmanijak reviewed (changes requested):
Hey, thanks for taking care of it! I tested the changes and found some issues but we're getting there!

One more thing is the Product Collection is left-aligned in Editor and centered in frontend when fixed width is applied

Editor | Frontend
-- | --
<img width="1099" alt="image" src="https://github.com/user-attachments/assets/f8af9a04-a99b-4be5-bbee-75ff9d2ca1b1"> | <img width="801" alt="Screenshot 2024-09-26 at 12 11 24" src="https://github.com/user-attachments/assets/99e2814d-4b71-4daf-a3d4-134ed4faaa53">

I don't have preferences in which way but they should be aligned 🤔 I guess centered makes more sense and brings more value until we add some alignment options.

octaedro committed:
Add line deleted by mistake

octaedro committed:
Fix editor alignment

octaedro committed:
Fix styles

octaedro commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -392,8 +434,7 @@ public function enhance_product_collection_with_interactivity( $block_content, $
 		}
-
-		return $block_content;
+		return sprintf( '<div>%s</div>', $block_content );
```
We add this `div` to center the product list on the front end when we select a percentage (%) `width`.

octaedro committed:
Add dimensions to templates and patterns

octaedro committed:
Make styles setting more defensive

octaedro committed:
Update plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/width-options-control.tsx

Co-authored-by: Karol Manijak <20098064+kmanijak@users.noreply.github.com>

octaedro commented:
Thank you @kmanijak for your review,

> I don't have preferences in which way but they should be aligned 🤔 I guess centered makes more sense and brings more value until we add some alignment options.

Good idea, I centered the items.

I addressed the changes you suggested and fixed the E2E tests. Could you take another look at this PR?

kmanijak committed:
Merge branch 'trunk' into add/47884_dimension_settings_to_product_collection

kmanijak reviewed (approved):
Thanks for addressing all of the comments, LGTM! Amazing work @octaedro! 🚀
